File: game/src/ai/Orientation.ts
// Orientation — the single source of truth for "which way is forward".
//
// THE BUG THIS FIXES: the rest of the AI historically derived a team's
// attacking ring from `match.faithEnd`. That is WRONG. Per the frozen match
// rules (src/match/MatchStateMachine.ts → attackDir):
//
//     home ALWAYS attacks the +X ring (x = +GATE_X)
//     away ALWAYS attacks the -X ring (x = -GATE_X)
//
// `faithEnd` only labels which physical ring scores as "Faith" (2 pts) vs
// "Free" (5 pts); it does NOT determine the direction a team drives the bell.
//
// Every nav target, receiver depth slot, throw lead, and Director posture must
// be expressed relative to THIS, never to faithEnd. This module is that anchor.

import type { MatchState, TeamSide, SimState } from '../sim/types';
import { GATE_X } from '../sim/RegConstants';

/** Signed attack direction for a team: +1 → drives toward +X ring, -1 → -X. */
export function attackSign(team: TeamSide): 1 | -1 {
  return team === 'home' ? 1 : -1;
}

/** World-x of the ring `team` is attacking (scoring at). */
export function attackRingX(team: TeamSide): number {
  return attackSign(team) * GATE_X;
}

/** World-x of the ring `team` is defending (their own ring). */
export function defendRingX(team: TeamSide): number {
  return -attackSign(team) * GATE_X;
}

/**
 * Is the ring `team` attacks the Faith (spinward, 2-pt) ring?
 * faithEnd '+x' means the +X ring is Faith. home attacks +X, away attacks -X.
 */
export function attackingFaithRing(team: TeamSide, match: MatchState): boolean {
  const ringX = attackRingX(team);
  const faithX = match.faithEnd === '+x' ? GATE_X : -GATE_X;
  return Math.sign(ringX) === Math.sign(faithX);
}

/**
 * Which team currently "owns the attack" — i.e. whose attacking direction the
 * offensive shape should be built around. Prefers the actual bell holder, then
 * the last thrower (in-flight from us), then match.possession.
 */
export function attackingTeam(state: SimState, match: MatchState): TeamSide {
  const heldBy = state.bell.heldBy;
  if (heldBy) {
    const h = state.players.find(p => p.id === heldBy);
    if (h) return h.team;
  }
  const thrownBy = state.bell.thrownBy;
  if (thrownBy && !state.bell.heldBy) {
    const t = state.players.find(p => p.id === thrownBy);
    if (t) return t.team;
  }
  return match.possession;
}

/** Forward (toward `team`'s attacking ring) progress of an x coordinate. */
export function forwardProgress(team: TeamSide, x: number): number {
  return attackSign(team) * x;
}

File: game/rig-core/src/ai/orientation.rs
//! 1:1 port of src/ai/Orientation.ts — the single source of truth for
//! "which way is forward". home ALWAYS attacks +X, away ALWAYS attacks
//! -X; faith_end only labels which ring scores Faith (2) vs Free (5),
//! never direction.

use super::types::{FaithEnd, MatchState, SimState, TeamSide};
use crate::tuning::GATE_X;

/// Signed attack direction: home → +1 (toward +X ring), away → -1.
pub fn attack_sign(team: TeamSide) -> f64 {
    match team {
        TeamSide::Home => 1.0,
        TeamSide::Away => -1.0,
    }
}

/// World-x of the ring `team` is attacking (scoring at).
pub fn attack_ring_x(team: TeamSide) -> f64 {
    attack_sign(team) * GATE_X
}

/// World-x of the ring `team` is defending (their own ring).
pub fn defend_ring_x(team: TeamSide) -> f64 {
    -attack_sign(team) * GATE_X
}

/// Is the ring `team` attacks the Faith (spinward, 2-pt) ring?
pub fn attacking_faith_ring(team: TeamSide, m: &MatchState) -> bool {
    let ring_x = attack_ring_x(team);
    let faith_x = match m.faith_end {
        FaithEnd::PlusX => GATE_X,
        FaithEnd::MinusX => -GATE_X,
    };
    ring_x.signum() == faith_x.signum()
}

/// Which team currently "owns the attack": bell holder, then last
/// thrower (in-flight from us), then match.possession.
pub fn attacking_team(state: &SimState, m: &MatchState) -> TeamSide {
    if let Some(held_by) = &state.bell.held_by {
        if let Some(h) = state.players.iter().find(|p| &p.id == held_by) {
            return h.team;
        }
    }
    if state.bell.held_by.is_none() {
        if let Some(thrown_by) = &state.bell.thrown_by {
            if let Some(t) = state.players.iter().find(|p| &p.id == thrown_by) {
                return t.team;
            }
        }
    }
    m.possession
}

/// Forward (toward `team`'s attacking ring) progress of an x coordinate.
pub fn forward_progress(team: TeamSide, x: f64) -> f64 {
    attack_sign(team) * x
}

/// OFFENSE REBUILD — signed world-x of a cast gate for `team` (mirrors
/// match_sm::gate_x: GATE_ABS = GATE_X·{0.25,0.55,0.85}, signed by attack
/// direction). The match advances the cast when a throw is SPENT (caught/
/// bobbled) with the bell at/past this line in the attack direction, so
/// the AI must drive + complete a pass past it to clear a gate.
pub fn gate_world_x(team: TeamSide, g: super::types::Gate) -> f64 {
    let frac = match g {
        super::types::Gate::First => 0.25,
        super::types::Gate::Deep => 0.55,
        super::types::Gate::Mouth => 0.85,
    };
    attack_sign(team) * GATE_X * frac
}
